armory/metrics/statistical.py

```diff
@@ -21,6 +22,54 @@ def register(metric):
     return set_namespace(registered, metric)
 
 
+@populationwise
+def precision_and_recall(y, y_pred):
+    """
+    Produce a dictionary whose keys are class labels, and values are (precision, recall) for that class
+    """
+    # Assumes that every class is represented in y
+
+    C = confusion_matrix(y, y_pred, normalize=False)
+    # breakpoint()
+    N = C.shape[0]
+    D = {}
+    for class_ in range(N):
+        # precision: true positives / number of items identified as class_
+        tp = C[class_, class_]
+        total_selected = C[:, class_].sum()
+        precision = tp / total_selected
+
+        # recall: true positives / number of actual items in class_
+        total_class_ = C[class_, :].sum()
+        recall = tp / total_class_
+
+        D[class_] = (precision, recall)
+
+    return D
+
+
+@populationwise
+def confusion_matrix(y, y_pred, normalize=True):
+    """
+    Produce a matrix C such that C[i,j] is the percentage of class i that was classified as j
+    If normalize is False, C[i,j] is the actual number of such elements, rather than the percentage
+    """
+    # Assumes that every class is represented in y
+    y = np.array(y)
+    y_pred = np.array(y_pred)
+    if y_pred.ndim == 2:
+        y_pred = np.argmax(y_pred, axis=1)
+    N = len(np.unique(y))
```

Comment:
If `y_pred` is 2D, you can use that to derive N. (Or at least check to ensure that they match).

Reply:
I may be misunderstanding, but N is the number of classes, not the total number of items.  Hence length of np.unique(y) and not length of y.  I don't think we can assume every class will show up in y_pred.  For that matter it seems a little risky to assume they will all be present in y.

Reply:
I think this implicitly assumes that the classes are all integers from 0 to N - 1. However, if y has missing classes, then there will be some misalignment.

Reply:
Oh right, of course.  Is it true that Armory scenarios will always have a 2D y_pred?  Or it just depends on how the meter and probes are set up?  So far the only source of a 1D y_pred I've encountered is my own unit tests, but I can expand those to 2D and then get N the way you described.

Reply:
Right now it's dependent on the underlying model, unfortunately.

Reply:
Well after all, if a class is totally absent from y, its row in the matrix would be all zeros since it was never classified as anything.  So maybe what I need to do is make this a dictionary after all, and key it with class labels, so if one is missing, then at least it will be clear what rows are what class.  Alternatively, I could add a row of zeros at the index of missing class labels, but this would only be possible for missing labels less than the greatest non-missing label.

Reply:
Let's just have the function assume that `y_pred` is 2D (and add that to the docstring). Other things can be handled by the user.